src/librustc_front/lowering.rs

```diff
@@ -1200,9 +1201,57 @@ pub fn lower_expr(lctx: &LoweringContext, e: &Expr) -> P<hir::Expr> {
             ExprIndex(ref el, ref er) => {
                 hir::ExprIndex(lower_expr(lctx, el), lower_expr(lctx, er))
             }
-            ExprRange(ref e1, ref e2) => {
-                hir::ExprRange(e1.as_ref().map(|x| lower_expr(lctx, x)),
-                               e2.as_ref().map(|x| lower_expr(lctx, x)))
+            ExprRange(ref e1, ref e2, lims) => {
+                fn make_struct(lctx: &LoweringContext,
+                               span: Span,
+                               path: Vec<&str>,
+                               fields: Vec<(&str, &P<Expr>)>) -> P<hir::Expr> {
+                    let strs = std_path(lctx, &iter::once("ops")
+                                                    .chain(path)
+                                                    .collect::<Vec<_>>());
+
+                    let structpath = path_global(span, strs);
+
+                    if fields.len() == 0 {
+                        expr_path(lctx,
+                                  structpath,
+                                  None)
+                    } else {
+                        expr_struct(lctx,
+                                    span,
+                                    structpath,
+                                    fields.into_iter().map(|(s, e)| {
+                                        field(token::intern(s), lower_expr(lctx, &**e), span)
+                                    }).collect(),
+                                    None,
+                                    None)
+                    }
+                }
+
+                return cache_ids(lctx, e.id, |lctx| {
+                    use syntax::ast::RangeLimits::*;
+                    make_struct(lctx,
+                                e.span,
+                                match (e1.is_some(), e2.is_some(), lims) {
```

Comment:
or better move the whole match into a local variable so it doesn't have so much rightward drift.